clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -17528,6 +17540,1901 @@ CodeGenFunction::EmitNVPTXBuiltinExpr(unsigned BuiltinID, const CallExpr *E) {
+  case NVPTX::BI__nvvm_atom_cta_acquire_inc_gen_ul:
+  case NVPTX::BI__nvvm_atom_cta_acquire_inc_gen_ull:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_inc_gen_i_cta_acquire);
+  case NVPTX::BI__nvvm_atom_cta_acquire_dec_gen_ui:
+  case NVPTX::BI__nvvm_atom_cta_acquire_dec_gen_ul:
+  case NVPTX::BI__nvvm_atom_cta_acquire_dec_gen_ull:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_dec_gen_i_cta_acquire);
+  case NVPTX::BI__nvvm_atom_sys_acquire_inc_gen_ui:
+  case NVPTX::BI__nvvm_atom_sys_acquire_inc_gen_ul:
+  case NVPTX::BI__nvvm_atom_sys_acquire_inc_gen_ull:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_inc_gen_i_sys_acquire);
+  case NVPTX::BI__nvvm_atom_sys_acquire_dec_gen_ui:
+  case NVPTX::BI__nvvm_atom_sys_acquire_dec_gen_ul:
+  case NVPTX::BI__nvvm_atom_sys_acquire_dec_gen_ull:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_dec_gen_i_sys_acquire);
+  case NVPTX::BI__nvvm_atom_cta_acquire_and_gen_i:
+  case NVPTX::BI__nvvm_atom_cta_acquire_and_gen_l:
+  case NVPTX::BI__nvvm_atom_cta_acquire_and_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_and_gen_i_cta_acquire);
+  case NVPTX::BI__nvvm_atom_sys_acquire_and_gen_i:
+  case NVPTX::BI__nvvm_atom_sys_acquire_and_gen_l:
+  case NVPTX::BI__nvvm_atom_sys_acquire_and_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_and_gen_i_sys_acquire);
+  case NVPTX::BI__nvvm_atom_cta_acquire_or_gen_i:
+  case NVPTX::BI__nvvm_atom_cta_acquire_or_gen_l:
+  case NVPTX::BI__nvvm_atom_cta_acquire_or_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_or_gen_i_cta_acquire);
+  case NVPTX::BI__nvvm_atom_sys_acquire_or_gen_i:
+  case NVPTX::BI__nvvm_atom_sys_acquire_or_gen_l:
+  case NVPTX::BI__nvvm_atom_sys_acquire_or_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_or_gen_i_sys_acquire);
+  case NVPTX::BI__nvvm_atom_cta_acquire_xor_gen_i:
+  case NVPTX::BI__nvvm_atom_cta_acquire_xor_gen_l:
+  case NVPTX::BI__nvvm_atom_cta_acquire_xor_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_xor_gen_i_cta_acquire);
+  case NVPTX::BI__nvvm_atom_sys_acquire_xor_gen_i:
+  case NVPTX::BI__nvvm_atom_sys_acquire_xor_gen_l:
+  case NVPTX::BI__nvvm_atom_sys_acquire_xor_gen_ll:
+    return MakeScopedAtomic(Intrinsic::nvvm_atomic_xor_gen_i_sys_acquire);
+  case NVPTX::BI__nvvm_atom_cta_acquire_cas_gen_i:
+  case NVPTX::BI__nvvm_atom_cta_acquire_cas_gen_l:
+  case NVPTX::BI__nvvm_atom_cta_acquire_cas_gen_ll: {
+    Value *Ptr = EmitScalarExpr(E->getArg(0));
+    return Builder.CreateCall(
+        CGM.getIntrinsic(
+            Intrinsic::nvvm_atomic_cas_gen_i_cta_acquire,
+            {Ptr->getType()->getPointerElementType(), Ptr->getType()}),
+        {Ptr, EmitScalarExpr(E->getArg(1)), EmitScalarExpr(E->getArg(2))});
+  }
+  case NVPTX::BI__nvvm_atom_sys_acquire_cas_gen_i:
+  case NVPTX::BI__nvvm_atom_sys_acquire_cas_gen_l:
+  case NVPTX::BI__nvvm_atom_sys_acquire_cas_gen_ll: {
+    Value *Ptr = EmitScalarExpr(E->getArg(0));
+    return Builder.CreateCall(
```

Comment:
Looks like this code is duplicated several times. Can this be factored out similar to `MakeScopedAtomic `

Reply:
Yes, done.

Reply:
Thanks!